modules/ding_place2book/ding_place2book.module

```diff
@@ -442,61 +442,65 @@ function ding_place2book_get_event_state($event_maker_id, $event_id) {
-        case 'published':
-          // Check if waiting list is activated.
-          switch ($meta_data->waiting_list) {
-            case 'true':
-              $status = 'waiting-list';
-              break;
-
-            default:
-              // Go through the sales status.
-              switch ($meta_data->sales_status) {
-                case 'open':
-                  // Check if there is any available tickets for sale.
-                  if ($meta_data->available_tickets > 0 || strval($meta_data->available_tickets) == '∞') {
-                    $status = 'open';
-                  }
-                  else {
-                    $status = 'sold-out';
-                  }
-                  break;
-
-                default:
-                  $status = 'closed';
-                  break;
-              }
-              break;
+    $end = strtotime($event->end_at);
+    $available_tickets = $meta_data->available_tickets;
+    $now = time();
+
+    // Try to derive an event status for the event by looking at the p2b event
+    // state, start date, available tickets and ticket types sales periods.
+    // See: http://developer.place2book.com/v1/events/ for a list possible p2b
+    // event states and what they mean.
+    $status = '';
+    // Start out by rulling out some easy conditions looking at the events start
+    // and end date.
+    if ($now >= $start && $now <= $end) {
+      // The event is currently taking place and ticket sales is over.
+      $status = 'closed';
+    }
+    elseif ($now > $end) {
+      // The event has already taking place.
+      $status = 'event-over';
+    }
+    elseif ($event_state === 'passive') {
+      $status = 'passive';
+    }
+    elseif ($event_state === 'published') {
+      // Now we know the event is not passive and is not over/taking place. If
+      // it's published it can now either be open, upcoming, closed or waiting
+      // list. Find out which by looking at the tickets and ticket types.
+      if ($available_tickets > 0 || strval($available_tickets) == '∞') {
+        // We have available tickets, so find out if sale is open or upcoming.
+        $prices = $p2b->getPrices($event_maker_id, $event_id);
+        foreach ($prices as $price) {
+          $price_sale_begin = strtotime($price->sale_begin_at);
+          $price_sale_end = strtotime($price->sale_end_at);
+          if ($now >= $price_sale_begin && $now <= $price_sale_end) {
+            // We have found an open sales period. No need to continue iterating
+            // the ticket types.
+            $status = 'open';
+            break;
           }
-          break;
-
-        case 'pause':
-          // Go through the sales status.
-          switch ($meta_data->sales_status) {
-            case 'upcoming':
-              $status = 'upcoming';
-              break;
-
-            default:
-              $status = 'closed';
-              break;
+          elseif ($now <= $price_sale_begin) {
```

Comment:
else clause missing

Reply:
But an else clause would break the logic? :)

Reply:
We only want to do something in these cases. Else just keep on iterating.

Reply:
Look at the code after the loop :)